mh/database.py

```diff
@@ -0,0 +1,328 @@
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+import random
+
+CHARSET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
+MEDIA_VERSIONS = {
+    "0903131810": "ROMJ08",
+    "0906222136": "RMHJ08",
+    "1001301045": "RMHE08",
+    "1002121503": "RMHP08",
+}
+
+
+def new_random_str(length=6):
+    return "".join(random.choice(CHARSET) for _ in range(length))
+
+
+class ServerType(object):
+    OPEN = 1
+    ROOKIE = 2
+    EXPERT = 3
+    RECRUITING = 4
+
+
+class LayerState(object):
+    JOINABLE = 0
+    EMPTY = 1
+    FULL = 2
+
+
+class City(object):
+    def __init__(self, name, parent):
+        self.name = name
+        self.parent = parent
+        self.capacity = 4
+        self.state = LayerState.EMPTY
+        self.players = set()
+
+    def get_population(self):
+        return len(self.players)
+
+    def get_capacity(self):
+        return self.capacity
+
+    def get_state(self):
+        size = self.get_population()
+        if size == 0:
+            return LayerState.EMPTY
+        elif size < self.get_capacity():
+            return LayerState.JOINABLE
+        else:
+            return LayerState.FULL
+
+
+class Gate(object):
+    def __init__(self, name, parent, city_count=40, player_capacity=100):
+        self.name = name
+        self.parent = parent
+        self.state = LayerState.EMPTY
+        self.capacity = player_capacity
+        self.cities = [
+            City("City{}".format(i), self)
+            for i in range(1, city_count+1)
+        ]
+        self.players = set()
+
+    def get_population(self):
+        return len(self.players) + sum((
+            city.get_population()
+            for city in self.cities
+        ))
+
+    def get_capacity(self):
+        return self.capacity
+
+    def get_state(self):
+        size = self.get_population()
+        if size == 0:
+            return LayerState.EMPTY
+        elif size < self.get_capacity():
+            return LayerState.JOINABLE
+        else:
+            return LayerState.FULL
+
+
+class Server(object):
+    def __init__(self, name, server_type, gate_count=40, capacity=2000,
+                 addr=None, port=None):
+        self.name = name
+        self.server_type = server_type
+        self.capacity = capacity
+        self.addr = addr
+        self.port = port
+        self.gates = [
+            Gate("City Gate{}".format(i), self)
+            for i in range(1, gate_count+1)
+        ]
+        self.players = set()
+
+    def get_population(self):
+        return len(self.players) + sum((
+            gate.get_population() for gate in self.gates
+        ))
+
+    def get_capacity(self):
+        return self.capacity
+
+
+def new_servers():
+    servers = []
```

Comment:
Should we start loading and configuring the servers/gate/cities from a config file?

Reply:
Wouldn't be better to start this part when we have a proper db schema and the necessary API for inter-server communication? Currently, we share the state across server using the TempDatabase.

Reply:
Sound reasonable